drivers/sensor/st/lsm6dsv16x/lsm6dsv16x_rtio_stream.c

```diff
@@ -0,0 +1,344 @@
+		lsm6dsv16x->streaming_sqe = NULL;
+		gpio_pin_interrupt_configure_dt(irq_gpio, GPIO_INT_EDGE_TO_ACTIVE);
+		return;
+	}
+
+	struct rtio_cqe *cqe;
+	int res = 0;
+
+	do {
+		cqe = rtio_cqe_consume(lsm6dsv16x->rtio_ctx);
+		if (cqe != NULL) {
+			if ((cqe->result < 0) && (res == 0)) {
+				LOG_ERR("Bus error: %d", cqe->result);
+				res = cqe->result;
+			}
+			rtio_cqe_release(lsm6dsv16x->rtio_ctx, cqe);
+		}
+	} while (cqe != NULL);
+
+	/* Bail/cancel attempt to read sensor on any error */
+	if (res != 0) {
+		rtio_iodev_sqe_err(lsm6dsv16x->streaming_sqe, res);
+		lsm6dsv16x->streaming_sqe = NULL;
+		return;
+	}
+
+	enum sensor_stream_data_opt data_opt;
+
+	if (has_fifo_ths_trig && !has_fifo_full_trig) {
+		/* Only care about fifo threshold */
+		data_opt = fifo_ths_cfg->opt;
+	} else if (!has_fifo_ths_trig && has_fifo_full_trig) {
+		/* Only care about fifo full */
+		data_opt = fifo_full_cfg->opt;
+	} else {
+		/* Both fifo threshold and full */
+		data_opt = MIN(fifo_ths_cfg->opt, fifo_full_cfg->opt);
+	}
+
+	if (data_opt == SENSOR_STREAM_DATA_NOP || data_opt == SENSOR_STREAM_DATA_DROP) {
+		uint8_t *buf;
+		uint32_t buf_len;
+
+		/* Clear streaming_sqe since we're done with the call */
+		if (rtio_sqe_rx_buf(lsm6dsv16x->streaming_sqe, sizeof(struct lsm6dsv16x_fifo_data),
+				    sizeof(struct lsm6dsv16x_fifo_data), &buf, &buf_len) != 0) {
+			rtio_iodev_sqe_err(lsm6dsv16x->streaming_sqe, -ENOMEM);
+			lsm6dsv16x->streaming_sqe = NULL;
+			gpio_pin_interrupt_configure_dt(irq_gpio, GPIO_INT_EDGE_TO_ACTIVE);
+			return;
+		}
+
+		struct lsm6dsv16x_fifo_data *rx_data = (struct lsm6dsv16x_fifo_data *)buf;
+
+		memset(buf, 0, buf_len);
+		rx_data->header.is_fifo = 1;
+		rx_data->header.timestamp = k_ticks_to_ns_floor64(k_uptime_ticks());
+		rx_data->int_status = lsm6dsv16x->fifo_status[1];
+		rx_data->fifo_count = 0;
+		rtio_iodev_sqe_ok(lsm6dsv16x->streaming_sqe, 0);
+		lsm6dsv16x->streaming_sqe = NULL;
+
+		if (data_opt == SENSOR_STREAM_DATA_DROP) {
+			const struct lsm6dsv16x_config *config = dev->config;
+			stmdev_ctx_t *ctx = (stmdev_ctx_t *)&config->ctx;
+
+			/* Flush the FIFO by disabling it. */
+			lsm6dsv16x_fifo_mode_set(ctx, LSM6DSV16X_BYPASS_MODE);
+			return;
+		}
+
+		gpio_pin_interrupt_configure_dt(irq_gpio, GPIO_INT_EDGE_TO_ACTIVE);
+		return;
+	}
+
+	uint8_t *buf, *read_buf;
+	uint32_t buf_len, buf_avail;
+	uint32_t req_len = LSM6DSV16X_FIFO_SIZE(fifo_count) + sizeof(struct lsm6dsv16x_fifo_data);
+
+	if (rtio_sqe_rx_buf(lsm6dsv16x->streaming_sqe, req_len, req_len, &buf, &buf_len) != 0) {
+		LOG_ERR("Failed to get buffer");
+		rtio_iodev_sqe_err(lsm6dsv16x->streaming_sqe, -ENOMEM);
+		lsm6dsv16x->streaming_sqe = NULL;
+		gpio_pin_interrupt_configure_dt(irq_gpio, GPIO_INT_EDGE_TO_ACTIVE);
+		return;
+	}
+
+	struct lsm6dsv16x_fifo_data hdr = {
+		.header = {
+			.is_fifo = true,
+			.accel_fs = lsm6dsv16x->accel_fs,
+			.gyro_fs = lsm6dsv16x->gyro_fs,
+			.timestamp = k_ticks_to_ns_floor64(k_uptime_ticks()),
```

Comment:
Shouldn't the timestamp be recorded when the interrupt occurs rather than when it gets done reading the fifo status

Reply:
Another way of  thinking about it is considering if you had an accelerometer running at 1hz and a FIFO threshold of 1 sample, and the sampling is aligned to N second boundaries.
```
T = 0.0 accelerometer start
T = 0.9 high priority task starts
T = 1.0 accelerometer samples, interrupt fires
T = 1.2 high priority task finishes
T = 1.3 read the FIFO status
```
The timestamp associated with the sample should not be 1.3, as reading the FIFO status is not strictly tied to the sampling.

Reply:
> Yes, like it is done in the rtio simple case. Good catch
> 
> EDIT: Well, actually it is before reading the FIFO. It is done just after reading the FIFO status and getting to know that it is actually a good interrupt. So, I think it is correct.

yes... but to get a accurate time stamp, it probably should record the "actual" time within that interrupt callback and save it in to the it's driver data and then write that recorded timestamp back in to the header once it realizes it's all good